File: ui/lib/api/web-tests.ts
/**
 * 版权所有 (c) 2023-2026 北京慧测信息技术有限公司(但问智能) 保留所有权利。
 * 
 * 本代码版权归北京慧测信息技术有限公司(但问智能)所有，仅用于学习交流目的，未经公司商业授权，
 * 不得用于任何商业用途，包括但不限于商业环境部署、售卖或以任何形式进行商业获利。违者必究。
 * 
 * 授权商业应用请联系微信：huice666
 */
// FIXME  MC80OmFIVnBZMlhsdktEbHVwNDZRM2h1WVE9PTo4OWFmMzFiMw==

/**
 * Web 测试相关的 API 客户端函数
 */

import { apiClient } from "./client";

// ==================== 类型定义 ====================

export interface WebTest {
  id: string;
  identifier: string;
  name: string;
  description?: string;
  base_url?: string;
  script_path: string;
  script_format: string;
  script_language: string;
  test_config?: Record<string, unknown>;
  target_pages?: string[];
  test_flows?: Record<string, unknown>[];
  total_pages: number;
  total_flows: number;
  folder_id?: string;
  created_at: string;
  updated_at?: string;
}

export interface WebTestRun {
  id: string;
  identifier: string;
  status: string;
  total_tests: number;
  passed_tests: number;
  failed_tests: number;
  skipped_tests: number;
  duration_ms?: number;
  error_message?: string;
  created_at: string;
}

export interface WebPage {
  id: string;
  display_name: string;
  url: string;
  page_title?: string;
  description?: string;
  page_type: string;
  elements?: Record<string, unknown>[];
  navigation?: Record<string, unknown>;
  tags?: string[];
  tag_group?: string;
  folder_id?: string;
  total_test_cases: number;
  total_test_runs: number;
  last_run_status?: string;
  created_at: string;
  updated_at?: string;
}

export interface CreateWebTestRequest {
  name: string;
  base_url: string;
  script_path: string;
  script_format?: string;
  script_language?: string;
  description?: string;
  test_config?: Record<string, unknown>;
  folder_id?: string;
  target_pages?: string[];
  test_flows?: Record<string, unknown>[];
}

// ==================== Web Test API 函数 ====================

export async function listWebTests(
  projectIdentifier: string,
  params?: {
    page?: number;
    page_size?: number;
    search?: string;
    script_format?: string;
  }
): Promise<{ items: WebTest[]; total: number; page: number; page_size: number }> {
  return apiClient.get(
    `/projects/${projectIdentifier}/web-tests`,
    { params }
  );
}

export async function getFolderWebTests(
  projectIdentifier: string,
  folderId: string | null,
  params?: {
    page?: number;
    page_size?: number;
  }
): Promise<{ items: WebTest[]; total: number; page: number; page_size: number }> {
  const url = folderId
    ? `/projects/${projectIdentifier}/web-tests/folder/${folderId}`
    : `/projects/${projectIdentifier}/web-tests`;

  return apiClient.get(url, { params });
}

export async function getWebTest(
  projectIdentifier: string,
  webTestId: string
): Promise<WebTest> {
  const response = await apiClient.get<{ data: WebTest }>(
    `/projects/${projectIdentifier}/web-tests/${webTestId}`
  );
  return response.data;
}
// eslint-disable  MS80OmFIVnBZMlhsdktEbHVwNDZRM2h1WVE9PTo4OWFmMzFiMw==

export async function createWebTest(
  projectIdentifier: string,
  data: CreateWebTestRequest
): Promise<WebTest> {
  const formData = new FormData();
  formData.append("name", data.name);
  formData.append("base_url", data.base_url);
  formData.append("script_path", data.script_path);
  formData.append("script_format", data.script_format || "playwright");
  formData.append("script_language", data.script_language || "typescript");

  if (data.description) {
    formData.append("description", data.description);
  }
  if (data.test_config) {
    formData.append("test_config", JSON.stringify(data.test_config));
  }
  if (data.folder_id) {
    formData.append("folder_id", data.folder_id);
  }
  if (data.target_pages) {
    formData.append("target_pages", JSON.stringify(data.target_pages));
  }
  if (data.test_flows) {
    formData.append("test_flows", JSON.stringify(data.test_flows));
  }

  const response = await fetch(`/api/v2/projects/${projectIdentifier}/web-tests`, {
    method: "POST",
    body: formData,
  });

  if (!response.ok) {
    const error = await response.json();
    throw new Error(error.detail || "Failed to create web test");
  }

  const result = await response.json();
  return result.data;
}

export async function updateWebTest(
  projectIdentifier: string,
  webTestId: string,
  data: {
    name?: string;
    description?: string;
    test_config?: Record<string, unknown>;
  }
): Promise<WebTest> {
  const response = await apiClient.patch<{ data: WebTest }>(
    `/projects/${projectIdentifier}/web-tests/${webTestId}`,
    data
  );
  return response.data;
}

export async function deleteWebTest(
  projectIdentifier: string,
  webTestId: string
): Promise<void> {
  await apiClient.delete(
    `/projects/${projectIdentifier}/web-tests/${webTestId}`
  );
}

export async function getWebTestScript(
  projectIdentifier: string,
  webTestId: string
): Promise<string> {
  const response = await fetch(
    `/api/v2/projects/${projectIdentifier}/web-tests/${webTestId}/script`
  );

  if (!response.ok) {
    throw new Error("Failed to get script");
  }

  return await response.text();
}

export async function updateWebTestScript(
  projectIdentifier: string,
  webTestId: string,
  scriptContent: string
): Promise<void> {
  const formData = new FormData();
  formData.append("script_content", scriptContent);

  await apiClient.put(
    `/projects/${projectIdentifier}/web-tests/${webTestId}/script`,
    formData as unknown as Record<string, unknown>
  );
}

export async function runWebTest(
  projectIdentifier: string,
  webTestId: string,
  executionConfig?: Record<string, unknown>
): Promise<{ run_id: string; status: string }> {
  const response = await apiClient.post<{ data: { run_id: string; status: string } }>(
    `/projects/${projectIdentifier}/web-tests/${webTestId}/run`,
    { execution_config: executionConfig }
  );
  return response.data;
}
// FIXME  Mi80OmFIVnBZMlhsdktEbHVwNDZRM2h1WVE9PTo4OWFmMzFiMw==

export async function getWebTestRuns(
  projectIdentifier: string,
  webTestId: string,
  params?: {
    page?: number;
    page_size?: number;
  }
): Promise<{ items: WebTestRun[]; total: number; page: number; page_size: number }> {
  return apiClient.get(
    `/projects/${projectIdentifier}/web-tests/${webTestId}/runs`,
    { params }
  );
}
// NOTE  My80OmFIVnBZMlhsdktEbHVwNDZRM2h1WVE9PTo4OWFmMzFiMw==

// ==================== Web Page API 函数 ====================

export async function listWebPages(
  projectIdentifier: string,
  params?: {
    page?: number;
    page_size?: number;
    folder_id?: string;
    search?: string;
  }
): Promise<{ items: WebPage[]; total: number; page: number; page_size: number }> {
  return apiClient.get(
    `/projects/${projectIdentifier}/web-pages`,
    { params }
  );
}

export async function getWebPage(
  pageId: string
): Promise<WebPage> {
  const response = await apiClient.get<{ data: WebPage }>(
    `/projects/web-pages/${pageId}`
  );
  return response.data;
}

export async function createWebPage(
  projectIdentifier: string,
  data: {
    display_name: string;
    url: string;
    folder_id?: string;
    page_title?: string;
    description?: string;
    page_type?: string;
    elements?: Record<string, unknown>[];
    navigation?: Record<string, unknown>;
    tags?: string[];
    tag_group?: string;
  }
): Promise<WebPage> {
  const response = await apiClient.post<{ data: WebPage }>(
    `/projects/${projectIdentifier}/web-pages`,
    data
  );
  return response.data;
}

export async function updateWebPage(
  pageId: string,
  data: {
    display_name?: string;
    url?: string;
    description?: string;
    elements?: Record<string, unknown>[];
    navigation?: Record<string, unknown>;
  }
): Promise<WebPage> {
  const response = await apiClient.patch<{ data: WebPage }>(
    `/projects/web-pages/${pageId}`,
    data
  );
  return response.data;
}

export async function deleteWebPage(
  pageId: string
): Promise<void> {
  await apiClient.delete(
    `/projects/web-pages/${pageId}`
  );
}

File: backend/app/api/v2/web_tests.py
"""
Web 测试管理 API

提供 Web 测试的 CRUD 操作接口
"""
"""
版权所有 (c) 2023-2026 北京慧测信息技术有限公司(但问智能) 保留所有权利。

本代码版权归北京慧测信息技术有限公司(但问智能)所有，仅用于学习交流目的，未经公司商业授权，
不得用于任何商业用途，包括但不限于商业环境部署、售卖或以任何形式进行商业获利。违者必究。

授权商业应用请联系微信：huice666
"""


from typing import Optional
from uuid import UUID

from fastapi import APIRouter, Query, UploadFile, File, Form, Depends
from fastapi.responses import Response, StreamingResponse

from app.api.deps import (
    PaginationDep,
)
from app.schemas.common import SuccessResponse
from app.services.web_test_service import WebTestService
from app.config.database import async_session_factory
from motor.motor_asyncio import AsyncIOMotorClient


router = APIRouter(prefix="/projects/{project_identifier}/web-tests")

# pylint: disable  MC80OmFIVnBZMlhsdktEbHVwNDZRMlJ5TXc9PTo4N2NjNzc1ZA==

async def get_web_test_service():
    """获取 Web 测试服务实例"""
    async with async_session_factory() as session:
        # mongodb = None  # 简化版本，不需要mongodb
        yield WebTestService(session, None)


# ============ Web 测试管理接口 ============

@router.post(
    "",
    response_model=SuccessResponse,
    summary="创建 Web 测试",
    description="创建新的 Web 测试",
)
async def create_web_test(
    project_identifier: str,
    service: WebTestService = Depends(get_web_test_service),
    name: str = Form(..., description="Web 测试名称"),
    base_url: str = Form(..., description="Web 应用基础 URL"),
    script_path: str = Form(..., description="脚本文件路径 (MinIO)"),
    script_format: str = Form(default="playwright", description="脚本格式"),
    script_language: str = Form(default="typescript", description="脚本语言"),
    description: Optional[str] = Form(None, description="描述"),
    test_config: Optional[str] = Form(None, description="测试配置 (JSON)"),
    folder_id: Optional[str] = Form(None, description="文件夹 ID"),
    target_pages: Optional[str] = Form(None, description="目标页面 (JSON)"),
    test_flows: Optional[str] = Form(None, description="测试流程 (JSON)"),
):
    """创建 Web 测试"""
    import json
    config = json.loads(test_config) if test_config else None
    pages = json.loads(target_pages) if target_pages else None
    flows = json.loads(test_flows) if test_flows else None

    result = await service.create_web_test(
        project_identifier=project_identifier,
        name=name,
        base_url=base_url,
        script_path=script_path,
        script_format=script_format,
        script_language=script_language,
        description=description,
        test_config=config,
        folder_id=folder_id,
        target_pages=pages,
        test_flows=flows,
    )
# pylint: disable  MS80OmFIVnBZMlhsdktEbHVwNDZRMlJ5TXc9PTo4N2NjNzc1ZA==

    return SuccessResponse(data=result, message="Web 测试创建成功")


@router.get(
    "",
    response_model=SuccessResponse,
    summary="获取 Web 测试列表",
    description="获取项目下的所有 Web 测试列表，支持搜索和过滤",
)
async def list_web_tests(
    project_identifier: str,
    service: WebTestService = Depends(get_web_test_service),
    pagination: PaginationDep = None,  # type: ignore
    search: Optional[str] = Query(None, description="搜索关键词"),
    script_format: Optional[str] = Query(None, description="脚本格式过滤"),
):
    """获取 Web 测试列表"""
    result = await service.list_web_tests(
        project_identifier=project_identifier,
        page=pagination.p,
        page_size=pagination.page_size,
        search=search,
        script_format=script_format,
    )

    return SuccessResponse(data=result)


@router.get(
    "/{web_test_id}",
    response_model=SuccessResponse,
    summary="获取 Web 测试详情",
    description="获取指定 Web 测试的详细信息",
)
async def get_web_test(
    project_identifier: str,
    web_test_id: str,
    service: WebTestService = Depends(get_web_test_service),
):
    """获取 Web 测试详情"""
    result = await service.get_web_test(
        project_identifier=project_identifier,
        web_test_id=web_test_id,
    )

    return SuccessResponse(data=result)


@router.patch(
    "/{web_test_id}",
    response_model=SuccessResponse,
    summary="更新 Web 测试",
    description="更新 Web 测试配置",
)
async def update_web_test(
    project_identifier: str,
    web_test_id: str,
    service: WebTestService = Depends(get_web_test_service),
    name: Optional[str] = None,
    description: Optional[str] = None,
    test_config: Optional[dict] = None,
):
    """更新 Web 测试"""
    result = await service.update_web_test(
        project_identifier=project_identifier,
        web_test_id=web_test_id,
        name=name,
        description=description,
        test_config=test_config,
    )

    return SuccessResponse(data=result, message="Web 测试更新成功")


@router.delete(
    "/{web_test_id}",
    response_model=SuccessResponse,
    summary="删除 Web 测试",
    description="删除指定的 Web 测试",
)
async def delete_web_test(
    project_identifier: str,
    web_test_id: str,
    service: WebTestService = Depends(get_web_test_service),
):
    """删除 Web 测试"""
    await service.delete_web_test(
        project_identifier=project_identifier,
        web_test_id=web_test_id,
    )

    return SuccessResponse(message="Web 测试删除成功")


# ============ 测试脚本管理接口 ============
# pragma: no cover  Mi80OmFIVnBZMlhsdktEbHVwNDZRMlJ5TXc9PTo4N2NjNzc1ZA==

@router.get(
    "/{web_test_id}/script",
    response_class=Response,
    summary="获取测试脚本",
    description="获取 Web 测试的脚本内容",
)
async def get_test_script(
    project_identifier: str,
    web_test_id: str,
    service: WebTestService = Depends(get_web_test_service),
):
    """获取测试脚本"""
    content = await service.get_test_script(
        project_identifier=project_identifier,
        web_test_id=web_test_id,
    )

    return Response(content=content, media_type="text/plain")


@router.put(
    "/{web_test_id}/script",
    response_model=SuccessResponse,
    summary="更新测试脚本",
    description="更新 Web 测试的脚本内容",
)
async def update_test_script(
    project_identifier: str,
    web_test_id: str,
    service: WebTestService = Depends(get_web_test_service),
    script_content: str = Form(..., description="脚本内容"),
):
    """更新测试脚本"""
    await service.update_test_script(
        project_identifier=project_identifier,
        web_test_id=web_test_id,
        script_content=script_content,
    )

    return SuccessResponse(message="测试脚本更新成功")

# pragma: no cover  My80OmFIVnBZMlhsdktEbHVwNDZRMlJ5TXc9PTo4N2NjNzc1ZA==

# ============ 测试执行接口 ============

@router.post(
    "/{web_test_id}/run",
    response_model=SuccessResponse,
    summary="执行 Web 测试",
    description="执行指定的 Web 测试",
)
async def run_web_test(
    project_identifier: str,
    web_test_id: str,
    service: WebTestService = Depends(get_web_test_service),
    execution_config: Optional[dict] = None,
):
    """执行 Web 测试"""
    result = await service.run_web_test(
        project_identifier=project_identifier,
        web_test_id=web_test_id,
        execution_config=execution_config,
    )

    return SuccessResponse(data=result, message="Web 测试已开始执行")


@router.get(
    "/{web_test_id}/runs",
    response_model=SuccessResponse,
    summary="获取测试运行历史",
    description="获取 Web 测试的运行历史记录",
)
async def list_test_runs(
    project_identifier: str,
    web_test_id: str,
    service: WebTestService = Depends(get_web_test_service),
    pagination: PaginationDep = None,  # type: ignore
):
    """获取测试运行历史"""
    result = await service.get_test_runs(
        project_identifier=project_identifier,
        web_test_id=web_test_id,
        page=pagination.p,
        page_size=pagination.page_size,
    )

    return SuccessResponse(data=result)


# ============ 文件夹操作接口 ============

@router.get(
    "/folder/{folder_id}",
    response_model=SuccessResponse,
    summary="获取文件夹下的 Web 测试",
    description="获取指定文件夹下的所有 Web 测试",
)
async def get_folder_web_tests(
    project_identifier: str,
    folder_id: str,
    service: WebTestService = Depends(get_web_test_service),
    pagination: PaginationDep = None,  # type: ignore
):
    """获取文件夹下的 Web 测试"""
    result = await service.get_folder_web_tests(
        project_identifier=project_identifier,
        folder_id=folder_id,
        page=pagination.p,
        page_size=pagination.page_size,
    )

    return SuccessResponse(data=result)
